language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,272 @@
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_verify!(self.invariant());
+        checked_postcondition!(!self.0.contains_key(&nonce));
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        checked_assume!(self.invariant());
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
+                    })
+                    .collect::<BTreeSet<_>>(),
+            );
+        }
+        let new_borrow_graph = BorrowGraph(new_graph);
+        checked_verify!(self.invariant());
+        checked_verify!(new_borrow_graph.invariant());
+        new_borrow_graph
+    }
+
+    /// checks if `self` covers `other`
+    pub fn abstracts(&self, other: &Self) -> bool {
+        self.unmatched_edges(other).values().all(|es| es.is_empty())
+    }
+
+    /// joins `other` into `self`
+    pub fn join(&mut self, other: &Self) {
+        for (n, es) in self.unmatched_edges(other) {
+            self.0.get_mut(&n).unwrap().extend(es);
+        }
+    }
+
+    /// gets all nonces that are targets of outgoing edges from `nonce`
+    pub fn all_borrows(&self, nonce: Nonce) -> BTreeSet<Nonce> {
+        checked_precondition!(self.0.contains_key(&nonce));
+        self.0[&nonce].iter().map(|x| x.to).collect()
+    }
+
+    /// gets all nonces that are targets of outgoing edges from `nonce` that are labeled with the empty label
+    pub fn nil_borrows(&self, nonce: Nonce) -> BTreeSet<Nonce> {
+        checked_precondition!(self.0.contains_key(&nonce));
+        self.0[&nonce]
+            .iter()
+            .filter(|x| x.label.is_empty())
+            .map(|x| x.to)
+            .collect()
+    }
+
+    /// gets all nonces that are targets of outgoing edges from `nonce` that are consistent with `label_elem`
+    pub fn consistent_borrows(&self, nonce: Nonce, label_elem: T) -> BTreeSet<Nonce> {
+        checked_precondition!(self.0.contains_key(&nonce));
+        self.0[&nonce]
+            .iter()
+            .filter(|x| x.label.is_empty() || x.label[0] == label_elem)
+            .map(|x| x.to)
+            .collect()
+    }
+
+    fn split<F>(edge_set: BTreeSet<Edge<T>>, pred: F) -> (Vec<Edge<T>>, BTreeSet<Edge<T>>)
+    where
+        F: Copy,
+        F: FnOnce(&Edge<T>) -> bool,
+    {
+        let mut pred_true = Vec::new();
+        let mut pred_false = BTreeSet::new();
+        for edge in edge_set.into_iter() {
```

Comment:
Once you have 
``` rust
let btrue_set: BTreeSet<_> = edge_set.iter().filter(pred).collect();
``` 
as another `BtreeSet`, you can just compute the remainder as `edge_set.sub(btrue_set)` because of its [Sub](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#impl-Sub%3C%26%27_%20BTreeSet%3CT%3E%3E) implementation. You'll return a set instead of a vector, but your'e doing the conversion to another BTreeSet in line 119 anyway.

Reply:
But won't this clone the entries satisfying pred when putting them into btrue_set?